src/AndersonAcceleration.f90

```diff
@@ -196,59 +143,217 @@ SUBROUTINE init_AndersonAccelerationType(solver,MPIEnv,Params)
   CLASS(AndersonAccelerationType),INTENT(INOUT) :: solver
-#ifdef FUTILITY_HAVE_Trilinos_NOX
-  IF(solver%iter==0) THEN
-    CALL Anderson_Reset(solver%id)
-  ELSE
-    CALL Anderson_Update(solver%id)
+  REAL(SRK),INTENT(IN) :: New_Gx(:)
+
+  INTEGER(SIK) :: i,j,k,depth_s
+  REAL(SRK) :: x_tmp(solver%N)
+
+  !Update iteration counter
+  solver%s=solver%s+1
+
+  IF(solver%s >= solver%start) THEN
+    depth_s=MIN(solver%depth,solver%s-solver%start)
+    !Push back vectors to make room for new ones
+    DO i=depth_s,1,-1
+      solver%Gx(:,i+1)=solver%Gx(:,i)
+      solver%r(:,i+1)=solver%r(:,i)
+    ENDDO
+
+    !Set new vectors based on input
+    solver%Gx(:,1)=New_Gx(:)
+    solver%r(:,1)=solver%Gx(:,1)-solver%x(:,1)
+
+    !Get fit coefficients
+    IF(depth_s == 1) THEN
+      solver%A(1,1)=0.0_SRK
+      solver%b(1)=0.0_SRK
+      DO j=1,solver%N
+        solver%A(1,1)=solver%A(1,1)+(solver%r(j,2)-solver%r(j,1))**2_SIK
+        solver%b(1)=solver%b(1)+solver%r(j,2)*(solver%r(j,2)-solver%r(j,1))
+      ENDDO
+      solver%alpha(1)=solver%b(1)/solver%A(1,1)
+    ELSEIF(depth_s > 1) THEN
+      !Construct coefficient matrix, right hand side, and solve for fit coefficients
+      DO i=1,depth_s
+        solver%b(i)=0.0_SRK
+        DO j=1,solver%N
+          solver%b(i)=solver%b(i)+solver%r(j,depth_s+1)*(solver%r(j,depth_s+1)-solver%r(j,i))
+        ENDDO
+        DO k=1,i
+          solver%A(k,i)=0.0_SRK
+          DO j=1,solver%N
+            solver%A(k,i)=solver%A(k,i)+(solver%r(j,depth_s+1)-solver%r(j,k))* &
+                                        (solver%r(j,depth_s+1)-solver%r(j,i))
+          ENDDO
+          IF(i /= k)solver%A(i,k)=solver%A(k,i)
+        ENDDO
+      ENDDO
+      solver%alpha(1:depth_s)=GECP(solver%A(1:depth_s,1:depth_s),solver%b(1:depth_s))
+    ENDIF
+    !Back out "0th" alpha
+    solver%alpha(depth_s+1)=0.0_SRK
+    DO i=1,depth_s
+      solver%alpha(depth_s+1)=solver%alpha(depth_s+1)+solver%alpha(i)
+    ENDDO
+    solver%alpha(depth_s+1)=1.0_SRK-solver%alpha(depth_s+1)
+
+    !Get accelerated solution
+    x_tmp(:)=0.0_SRK
+    DO i=1,depth_s+1
+      DO j=1,solver%N
+        x_tmp(j)=x_tmp(j)+solver%alpha(i)*(solver%x(j,i)+solver%beta*solver%r(j,i))
+      ENDDO
+    ENDDO
+
+    !Push back solution vectors and load in newest accelerated as next initial iterate
+    DO i=MIN(depth_s+1,solver%depth),1,-1
+      solver%x(:,i+1)=solver%x(:,i)
+    ENDDO
+    solver%x(:,1)=x_tmp(:)
   ENDIF
 
-  solver%iter=solver%iter+1
-#endif
-ENDSUBROUTINE step_AndersonAccelerationType
+ENDFUNCTION step_AndersonAccelerationType
 !
 !-------------------------------------------------------------------------------
-!> @brief Reset the Anderson acceleration solver
+!> @brief Set the initial iterate for the Anderson acceleration solver
 !> @param solver The anderson solver to act on
+!> @param x initial iterate
 !>
-!> This routine resets the counters for the anderson acceleration
-!>
-SUBROUTINE reset_AndersonAccelerationType(solver)
```

Comment:
still think this should be reset.  maybe you need both but they do the same thing.  I'm fine with that.  Just something that says, this is new, clear the history and start over.

Reply:
You're absolutely right about this. Needs to be reset for mutli-state problems.